src/WebJobs.Script/Rpc/LanguageWorkerChannelManager.cs

```diff
@@ -164,50 +175,57 @@ internal void ScheduleShutdownStandbyChannels()
+                        }
+                    }
                 }
             }
         }
 
-        public void ShutdownStandbyChannels(IEnumerable<FunctionMetadata> functions)
+        public void ShutdownChannels()
         {
-            if (_environment.GetEnvironmentVariable(EnvironmentSettingNames.AzureWebsitePlaceholderMode) == "1")
+            foreach (string runtime in _workerChannels.Keys)
             {
-                return;
-            }
-            _workerRuntime = _environment.GetEnvironmentVariable(LanguageWorkerConstants.FunctionWorkerRuntimeSettingName) ?? Utility.GetWorkerRuntime(functions);
-            _logger.LogInformation("WorkerRuntime: {workerRuntime}. Will shutdown other standby channels", _workerRuntime);
-            if (string.IsNullOrEmpty(_workerRuntime))
-            {
-                ShutdownChannels();
-                return;
-            }
-            else
-            {
-                ScheduleShutdownStandbyChannels();
+                _logger.LogInformation("Shutting down language worker channels for runtime:{runtime}", runtime);
+                foreach (var channel in _workerChannels[runtime].ToList())
+                {
+                    if (channel != null)
+                    {
+                        channel.Dispose();
+                        _workerChannels[runtime].Remove(channel);
+                    }
+                }
             }
         }
 
-        public void ShutdownChannels()
+        public void ShutdownProcessRegistry()
         {
-            foreach (string runtime in _workerChannels.Keys.ToList())
-            {
-                _logger.LogInformation("Shutting down language worker channel for runtime:{runtime}", runtime);
-                _workerChannels[runtime]?.Dispose();
-            }
-            _workerChannels.Clear();
-            (_processRegistry as IDisposable)?.Dispose();
+           (_processRegistry as IDisposable)?.Dispose();
         }
 
         private void AddOrUpdateWorkerChannels(RpcWebHostChannelReadyEvent rpcChannelReadyEvent)
         {
-            _logger.LogInformation("Adding language worker channel for runtime: {language}.", rpcChannelReadyEvent.Language);
-            _workerChannels.Add(rpcChannelReadyEvent.Language, rpcChannelReadyEvent.LanguageWorkerChannel);
+            _logger.LogDebug("Adding language worker channel for runtime: {language}.", rpcChannelReadyEvent.Language);
+            List<ILanguageWorkerChannel> languageWorkerChannels = null;
+            if (_workerChannels.TryGetValue(rpcChannelReadyEvent.Language, out languageWorkerChannels))
+            {
+                languageWorkerChannels.Add(rpcChannelReadyEvent.LanguageWorkerChannel);
+            }
+            else
+            {
+                languageWorkerChannels = new List<ILanguageWorkerChannel>();
+                _workerChannels.TryAdd(rpcChannelReadyEvent.Language, languageWorkerChannels);
```

Comment:
You may need to evaluate whether the list was added to the dictionary. If this is called concurrently, there is a chance another thread will hit this first and your add below will be done against the wrong instance.

Reply:
`if` block above is checking if there already a list of channels for that language. Any better way of doing this? should there be lock?

Reply:
No, you just need to check the result of your `TryAdd` call, if it returns false, the list you're using is not the one that was added to the dictionary

Reply:
To be sane, you can also just add a new list and retrieve it after instead of using a local